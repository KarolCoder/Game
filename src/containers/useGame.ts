import {
  GameCategory,
  GameResource,
  GameValues,
  GameResult,
  ResourcesFromData,
} from '@/utils/types';
import {useState, useCallback} from 'react';
import {createContainer} from 'unstated-next';

const useGameBase = () => {
  const [selectedGameValues, setSelectedGameValues] = useState<GameValues>();

  const updateGameCategory = useCallback((category?: GameCategory) => {
    setSelectedGameValues(prev => ({...prev, category}));
  }, []);

  const updateGameResource = useCallback((resource?: GameResource) => {
    setSelectedGameValues(prev => ({...prev, resource}));
  }, []);

  const updateGameSelectedCard = useCallback((selectedCard?: number) => {
    setSelectedGameValues(prev => ({...prev, selectedCard}));
  }, []);

  const updateGameWinCounter = useCallback((winner: 'You' | 'Opponent') => {
    setSelectedGameValues(prev => ({
      ...prev,
      winCounter:
        winner === 'You'
          ? {
              opponent: prev?.winCounter?.opponent || 0,
              you: (prev?.winCounter?.you || 0) + 1,
            }
          : {
              opponent: (prev?.winCounter?.opponent || 0) + 1,
              you: prev?.winCounter?.you || 0,
            },
    }));
  }, []);

  const updateGameResult = useCallback((result?: GameResult) => {
    setSelectedGameValues(prev => ({...prev, result}));
  }, []);

  const updateGameResourcesFromData = useCallback(
    (resourcesFromData?: ResourcesFromData) => {
      setSelectedGameValues(prev => ({...prev, resourcesFromData}));
    },
    [],
  );

  const resetGameValues = useCallback(() => {
    setSelectedGameValues(undefined);
  }, []);

  return {
    selectedGameValues,
    updateGameCategory,
    updateGameResource,
    resetGameValues,
    updateGameSelectedCard,
    updateGameWinCounter,
    updateGameResult,
    updateGameResourcesFromData,
  };
};

const useGameContainer = createContainer(useGameBase);
export const useGame = useGameContainer.useContainer;
export const GameProvider = useGameContainer.Provider;
