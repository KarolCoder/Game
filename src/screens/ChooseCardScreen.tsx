import {PeopleResurcesUri, StarshipsResourcesUri} from '@/api/types';
import {Spacer} from '@/components';
import {useFetch} from '@/hooks/useFetch';
import {MainParamList} from '@/navigation/types/MainParamList';
import {getRandomInt, getResult} from '@/utils/functions';
import {RouteProp, useNavigation, useRoute} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import React, {useEffect, useState} from 'react';
import {
  Paragraph,
  Title,
  useTheme,
  Button,
  ActivityIndicator,
} from 'react-native-paper';
import styled from 'styled-components/native';

export const ChooseCardScreen = () => {
  const {colors} = useTheme();
  const {
    params: {activeCategory, activeResource},
  } = useRoute<RouteProp<MainParamList, 'ChooseCard'>>();
  const {navigate} = useNavigation<NativeStackNavigationProp<MainParamList>>();
  const {data, status, refreshData, resetData, getNextPage, nextPage} =
    useFetch<PeopleResurcesUri | StarshipsResourcesUri>({
      url: activeCategory === 'People' ? 'people' : 'starships',
      waitForFetching: true,
    });
  const [selectedCard, setSelectedCard] = useState<number | undefined>();
  const [resourcesFromData, setResourcesFromData] = useState<{
    firstResource: string | string[] | undefined;
    secondResource: string | string[] | undefined;
  }>();
  const [result, setResult] = useState<
    'Tie' | 'Winner' | 'Loser' | 'unknown' | undefined
  >('unknown');
  const [winCounter, setWinCounter] =
    useState<{you: number; opponent: number}>();

  useEffect(() => {
    const resourceFromDataFirst =
      data && data.results && data.results.length > 0
        ? data.results[getRandomInt(0, data.results.length - 1)][
            activeResource.uriParam
          ]
        : undefined;
    const resourceFromDataSecond =
      data && data.results && data.results.length > 0
        ? data.results[getRandomInt(0, data.results.length - 1)][
            activeResource.uriParam
          ]
        : undefined;
    setResourcesFromData({
      firstResource: resourceFromDataFirst,
      secondResource: resourceFromDataSecond,
    });
    const dataResult = data
      ? getResult({
          firstValue: resourceFromDataFirst,
          secondValue: resourceFromDataSecond,
          resource: activeResource.uriParam,
          selectedCard: selectedCard,
        })
      : 'unknown';

    if (dataResult === 'Winner') {
      setWinCounter(prev => {
        if (!prev) {
          return {you: 1, opponent: 0};
        } else {
          return {...prev, you: prev.you + 1};
        }
      });
    } else if (dataResult === 'Loser') {
      setWinCounter(prev => {
        if (!prev) {
          return {you: 0, opponent: 1};
        } else {
          return {...prev, opponent: prev.opponent + 1};
        }
      });
    }
    setResult(dataResult);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data]);

  const firstValueDisplayed = Array.isArray(resourcesFromData?.firstResource)
    ? resourcesFromData?.firstResource.length.toString()
    : resourcesFromData?.firstResource;

  const secondValueDisplayed = Array.isArray(resourcesFromData?.secondResource)
    ? resourcesFromData?.secondResource.length.toString()
    : resourcesFromData?.secondResource;

  return (
    <Main backgroundColor={colors.background}>
      {status === 'fetching' && (
        <ActivityIndicatorWrapper>
          <ActivityIndicator animating size={'large'} />
        </ActivityIndicatorWrapper>
      )}
      <Row>
        <CategoryBox>
          <SmallText textColor={colors.primary}>Category</SmallText>
          <BigText textColor={colors.primary}>{activeCategory}</BigText>
        </CategoryBox>
        <CategoryBox>
          <SmallText textColor={colors.primary}>Resource</SmallText>
          <BigText textColor={colors.primary}>{activeResource.message}</BigText>
        </CategoryBox>
      </Row>
      <SelectTitle textColor={colors.primary}>Choose card</SelectTitle>
      <Row>
        <Card
          disabled={
            (!!resourcesFromData?.firstResource &&
              !!resourcesFromData?.secondResource) ||
            status === 'fetching'
          }
          onPress={() => {
            setSelectedCard(0);
            if (!nextPage) {
              refreshData();
            } else {
              getNextPage();
            }
          }}>
          <Title>{firstValueDisplayed ? firstValueDisplayed : '?'}</Title>
        </Card>
        <Spacer width={24} />
        <Card
          disabled={
            (!!resourcesFromData?.firstResource &&
              !!resourcesFromData?.secondResource) ||
            status === 'fetching'
          }
          onPress={() => {
            setSelectedCard(1);
            if (!nextPage) {
              refreshData();
            } else {
              getNextPage();
            }
          }}>
          <Title>{secondValueDisplayed ? secondValueDisplayed : '?'}</Title>
        </Card>
      </Row>
      {status === 'error' && (
        <SmallText textColor={colors.error}>
          There was some issue, please try again
        </SmallText>
      )}
      {result === 'Winner' ? (
        <BigText textColor={colors.primary}>
          Congratualtions, you won!!!
        </BigText>
      ) : result === 'Loser' ? (
        <BigText textColor={colors.primary}>You lose, loser :D</BigText>
      ) : result === 'Tie' ? (
        <BigText textColor={colors.primary}>Tie, try again</BigText>
      ) : !result ? (
        <BigText textColor={colors.primary}>
          Couldn't resolve winner please go to start
        </BigText>
      ) : null}
      {winCounter && (
        <Row>
          <CategoryBox>
            <SmallText textColor={colors.primary}>Your wins</SmallText>
            <BigText textColor={colors.primary}>{winCounter.you}</BigText>
          </CategoryBox>
          <CategoryBox>
            <SmallText textColor={colors.primary}>Opponent wins</SmallText>
            <BigText textColor={colors.primary}>{winCounter.opponent}</BigText>
          </CategoryBox>
        </Row>
      )}
      {!!resourcesFromData?.firstResource &&
        !!resourcesFromData?.secondResource && (
          <Row>
            <BottomButton
              mode="contained"
              disabled={!(activeResource && activeCategory)}
              onPress={() => resetData()}>
              Reset cards
            </BottomButton>
            <Spacer width={24} />
            <BottomButton
              mode="contained"
              disabled={!(activeResource && activeCategory)}
              onPress={() => navigate('Welcome')}>
              Go to start
            </BottomButton>
          </Row>
        )}
    </Main>
  );
};

const Main = styled.ScrollView.attrs(() => ({
  contentContainerStyle: {
    flexGrow: 1,
    alignItems: 'center',
    paddingBottom: 16,
    paddingHorizontal: 16,
  },
}))<{backgroundColor: string}>`
  flex: 1;
  background-color: ${({backgroundColor}) => backgroundColor};
`;

const ActivityIndicatorWrapper = styled.View`
  position: absolute;
  align-items: center;
  justify-content: center;
  top: 0px;
  bottom: 0px;
  left: 0px;
  right: 0px;
  z-index: 1;
`;

const CategoryBox = styled.View`
  justify-content: center;
  align-items: center;
`;

const BigText = styled(Title)<{textColor: string}>`
  color: ${({textColor}) => textColor};
`;

const SmallText = styled(Paragraph)<{textColor: string}>`
  color: ${({textColor}) => textColor};
`;

const SelectTitle = styled(Title)<{textColor: string}>`
  margin: 24px 0px 12px;
  color: ${({textColor}) => textColor};
`;
const Row = styled.View`
  flex-direction: row;
  justify-content: space-evenly;
  width: 100%;
  padding: 24px 16px;
`;

const Card = styled.TouchableOpacity`
  justify-content: center;
  align-items: center;
  background-color: white;
  aspect-ratio: 0.6;
  flex: 1;
  border-radius: 8px;
`;

const BottomButton = styled(Button)`
  padding: 4px 8px;
  flex: 1;
`;
