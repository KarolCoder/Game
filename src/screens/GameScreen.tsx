import {
  peopleResourcesToBattle,
  starshipsResourcesToBattle,
} from '@/api/consts';
import {PeopleResurcesUri, Resource, StarshipsResourcesUri} from '@/api/types';
import {Spacer} from '@/components';
import {MainParamList} from '@/navigation/types/MainParamList';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import React, {useState} from 'react';
import {useEffect} from 'react';
import {Button, Title, useTheme} from 'react-native-paper';
import styled, {css} from 'styled-components/native';

export const GameScreen = () => {
  const {colors} = useTheme();
  const {navigate} = useNavigation<NativeStackNavigationProp<MainParamList>>();
  const [activeCategory, setActiveCategory] = useState<
    'Starships' | 'People'
  >();
  const [activeResource, setActiveResource] =
    useState<Resource<StarshipsResourcesUri | PeopleResurcesUri>>();

  useEffect(() => {
    setActiveResource(undefined);
  }, [activeCategory]);

  const resources =
    activeCategory === 'People'
      ? peopleResourcesToBattle
      : activeCategory === 'Starships'
      ? starshipsResourcesToBattle
      : [];

  return (
    <Main backgroundColor={colors.background}>
      <SelectTitle textColor={colors.primary}>Choose category</SelectTitle>
      <Row>
        <OptionButton
          isActive={activeCategory === 'People'}
          mode="contained"
          onPress={() => setActiveCategory('People')}>
          People
        </OptionButton>
        <Spacer width={24} />
        <OptionButton
          isActive={activeCategory === 'Starships'}
          mode="contained"
          onPress={() => setActiveCategory('Starships')}>
          Starships
        </OptionButton>
      </Row>
      {activeCategory && (
        <SelectTitle textColor={colors.primary}>Choose resource</SelectTitle>
      )}
      <ResourcesList>
        {resources.map(({message, uriParam}) => (
          <ResourceButton
            key={message}
            mode="contained"
            isActive={message === activeResource?.message}
            onPress={() => setActiveResource({message, uriParam})}>
            {message}
          </ResourceButton>
        ))}
      </ResourcesList>
      <ContinueButton
        mode="contained"
        disabled={!(activeResource && activeCategory)}
        onPress={() => {
          if (activeCategory && activeResource) {
            navigate('ChooseCard', {
              activeCategory: activeCategory,
              activeResource: activeResource,
            });
          }
        }}>
        Continue
      </ContinueButton>
    </Main>
  );
};

export const Main = styled.View<{backgroundColor: string}>`
  flex: 1;
  background-color: ${({backgroundColor}) => backgroundColor};
  justify-content: center;
  align-items: center;
`;

const SelectTitle = styled(Title)<{textColor: string}>`
  margin: 24px 0px 12px;
  color: ${({textColor}) => textColor};
`;

export const Row = styled.View`
  flex-direction: row;
  width: 100%;
  padding: 0px 16px;
`;

const OptionButton = styled(Button)<{isActive?: boolean}>`
  padding: ${({isActive}) => (isActive ? '0px 4px' : '4px')};
  flex: 1;
  justify-content: center;
  ${({isActive}) =>
    isActive &&
    css`
      border: 4px solid #ffffff;
    `}
`;

const ResourcesList = styled.ScrollView.attrs(() => ({
  contentContainerStyle: {
    flexGrow: 1,
    paddingBottom: 16,
    paddingHorizontal: 16,
  },
}))`
  flex: 1;
  width: 100%;
`;

const ResourceButton = styled(Button)<{isActive?: boolean}>`
  padding: ${({isActive}) => (isActive ? '0px 4px' : '4px')};
  margin-bottom: 8px;
  ${({isActive}) =>
    isActive &&
    css`
      border: 4px solid #ffffff;
    `}
`;

const ContinueButton = styled(Button)<{disabled?: boolean}>`
  padding: 4px 8px;
  margin-bottom: 16px;
  margin-top: 24px;
  background-color: ${({disabled}) => (disabled ? '	#E8E8E855' : '#F8F8F8')};
`;
