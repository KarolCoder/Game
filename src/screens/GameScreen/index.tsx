import {MainParamList} from '@/navigation/types/MainParamList';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import React from 'react';
import {useTheme} from 'react-native-paper';
import {useGameScreen} from './hooks';
import {CategorySection} from './partials/CategorySection';
import {
  ContinueButton,
  Main,
  ResourceButton,
  ResourcesList,
  SelectTitle,
} from './styles';

export const GameScreen = () => {
  const {colors} = useTheme();
  const {navigate} = useNavigation<NativeStackNavigationProp<MainParamList>>();
  const {
    resources,
    activeResource,
    setActiveCategory,
    activeCategory,
    setActiveResource,
  } = useGameScreen();

  return (
    <Main backgroundColor={colors.background}>
      <CategorySection
        activeCategory={activeCategory}
        setActiveCategory={setActiveCategory}
      />
      {activeCategory && (
        <SelectTitle textColor={colors.primary}>Choose resource</SelectTitle>
      )}
      <ResourcesList>
        {resources.map(({message, uriParam}) => (
          <ResourceButton
            key={message}
            mode="contained"
            isActive={message === activeResource?.message}
            onPress={() => setActiveResource({message, uriParam})}>
            {message}
          </ResourceButton>
        ))}
      </ResourcesList>
      <ContinueButton
        mode="contained"
        disabled={!(activeResource && activeCategory)}
        onPress={() => {
          if (activeCategory && activeResource) {
            navigate('ChooseCard', {
              activeCategory: activeCategory,
              activeResource: activeResource,
            });
          }
        }}>
        Continue
      </ContinueButton>
    </Main>
  );
};
