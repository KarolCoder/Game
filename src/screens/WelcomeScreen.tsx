import {MainParamList} from '@/navigation/types/MainParamList';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import React from 'react';

import {Button, Title, useTheme} from 'react-native-paper';
import styled from 'styled-components/native';

export const WelcomeScreen = () => {
  const {colors} = useTheme();
  const {navigate} = useNavigation<NativeStackNavigationProp<MainParamList>>();
  return (
    <StarWarsImage
      source={require('@/assets/images/starWars.jpeg')}
      resizeMode="cover">
      <TopText textColor={colors.primary}>Star Wars Game</TopText>
      <ButtonWrapper>
        <BottomButton mode="contained" onPress={() => navigate('Game')}>
          <Title>Play</Title>
        </BottomButton>
      </ButtonWrapper>
    </StarWarsImage>
  );
};

const StarWarsImage = styled.ImageBackground`
  flex: 1;
  align-items: center;
  width: 100%;
  height: 100%;
`;

const ButtonWrapper = styled.View`
  width: 100%;
  margin-top: auto;
  padding: 0px 16px;
`;

const BottomButton = styled(Button)`
  margin-bottom: 16px;
`;

const TopText = styled(Title)<{textColor: string}>`
  margin-top: 40px;
  color: ${({textColor}) => textColor};
`;
