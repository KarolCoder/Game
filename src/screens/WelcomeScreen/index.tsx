import {MainParamList} from '@/navigation/types/MainParamList';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import React from 'react';

import {Title, useTheme} from 'react-native-paper';
import {BottomButton, ButtonWrapper, StarWarsImage, TopText} from './styles';

export const WelcomeScreen = () => {
  const {colors} = useTheme();
  const {navigate} = useNavigation<NativeStackNavigationProp<MainParamList>>();
  return (
    <StarWarsImage
      source={require('@/assets/images/starWars.jpeg')}
      resizeMode="cover">
      <TopText textColor={colors.primary}>Star Wars Game</TopText>
      <ButtonWrapper>
        <BottomButton
          accessibilityHint="bottom button"
          mode="contained"
          onPress={() => navigate('Game')}>
          <Title>Play</Title>
        </BottomButton>
      </ButtonWrapper>
    </StarWarsImage>
  );
};
