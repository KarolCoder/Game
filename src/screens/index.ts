export {GameScreen} from './GameScreen';
export {WelcomeScreen} from './WelcomeScreen';
export {ChooseCardScreen} from './ChooseCardScreen';
