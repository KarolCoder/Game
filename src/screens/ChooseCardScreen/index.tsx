import {useGame} from '@/containers/useGame';
import {MainParamList} from '@/navigation/types/MainParamList';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import React from 'react';
import {useTheme, ActivityIndicator} from 'react-native-paper';
import {useChooseCardScreen} from './hooks';
import {BottomButtons} from './partials/BottomButtons';
import {CardsSection} from './partials/CardsSection';
import {ResultLabel} from './partials/ResultLabel';
import {TextSection} from './partials/TextSection';
import {ActivityIndicatorWrapper, Main, SmallText} from './styles';

export const ChooseCardScreen = () => {
  const {colors} = useTheme();
  const {navigate} = useNavigation<NativeStackNavigationProp<MainParamList>>();
  const {
    firstValueDisplayed,
    secondValueDisplayed,
    status,
    refreshData,
    resetData,
    getNextPage,
    nextPage,
    winCounter,
    result,
    resourcesFromData,
  } = useChooseCardScreen();
  const {updateGameSelectedCard, selectedGameValues} = useGame();
  const {category, resource} = {...selectedGameValues};

  return (
    <Main backgroundColor={colors.background}>
      {status === 'fetching' && (
        <ActivityIndicatorWrapper>
          <ActivityIndicator animating size={'large'} />
        </ActivityIndicatorWrapper>
      )}
      <TextSection
        leftTopText="Category"
        bottomLeftText={category}
        rightTopText="Resource"
        bottomRightText={resource?.message}
      />
      <CardsSection
        {...{
          resourcesFromData,
          updateGameSelectedCard,
          refreshData,
          getNextPage,
          firstValueDisplayed,
          secondValueDisplayed,
          nextPage,
          status,
        }}
      />
      {status === 'error' && (
        <SmallText textColor={colors.error}>
          There was some issue, please try again
        </SmallText>
      )}
      <ResultLabel result={result} />
      {winCounter && (
        <TextSection
          leftTopText="Your wins"
          bottomLeftText={winCounter.you}
          rightTopText="Opponent wins"
          bottomRightText={winCounter.opponent}
        />
      )}
      {!!resourcesFromData?.firstResource &&
        !!resourcesFromData?.secondResource && (
          <BottomButtons
            onLeftPress={() => resetData()}
            onRightPress={() => navigate('Welcome')}
            isDisabled={!(resource && category)}
          />
        )}
    </Main>
  );
};
