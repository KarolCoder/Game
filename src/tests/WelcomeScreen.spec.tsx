import {GameProvider} from '@/containers/useGame';
import {WelcomeScreen} from '@/screens';
import React from 'react';

import renderer, {act} from 'react-test-renderer';

it('welcome screen renders correctly', async () => {
  const result = renderer
    .create(
      <GameProvider>
        <WelcomeScreen />
      </GameProvider>,
    )
    .toJSON();
  await act(async () => {
    expect(result).toMatchSnapshot();
    return;
  });
});
