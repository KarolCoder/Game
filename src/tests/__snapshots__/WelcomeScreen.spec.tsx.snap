// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`welcome screen renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "testUri": "../../../src/assets/images/starWars.jpeg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        Array [
          Object {
            "textAlign": "left",
          },
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "System",
            "fontWeight": "500",
            "writingDirection": "ltr",
          },
          Array [
            Object {
              "fontSize": 20,
              "letterSpacing": 0.15,
              "lineHeight": 30,
              "marginVertical": 2,
            },
            Array [
              Object {
                "color": "#6200ee",
                "marginTop": 40,
              },
            ],
          ],
        ],
      ]
    }
    textColor="#6200ee"
  >
    Star Wars Game
  </Text>
  <View
    style={
      Array [
        Object {
          "marginTop": "auto",
          "paddingBottom": 0,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 0,
          "width": "100%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#6200ee",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "marginBottom": 16,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 4,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "500",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 20,
                        "letterSpacing": 0.15,
                        "lineHeight": 30,
                        "marginVertical": 2,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
            >
              Play
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
